lib/tz_world.ex

```diff
@@ -1,89 +1,139 @@
+
+      iex> TzWorld.timezone_at(3.2, 45.32)
+      {:ok, "Europe/Paris"}
+
+
+  The algorithm starts by filtering out timezones whose bounding
+  box does not contain the given point.
+
+  Once filtered, the first timezone which contains the given
+  point is returned, or `nil` if none of the timzones match.
 
-  The algorithm starts by filtering out timezones whose bounding box does not contain
-  the given point.
-  Once filtered, we return the first timezone which contains the given point, or `nil` if
-  none of the timzones match.
   """
-  @spec timezone_at(Geo.Point.t) :: String.t | nil
+  @spec timezone_at(Geo.Point.t()) :: String.t() | nil
   def timezone_at(%Point{} = point) do
-    GenServer.call(__MODULE__, {:timezone_at, point})
+    GenServer.call(__MODULE__, {:timezone_at, point}, @timeout)
   end
 
-  def handle_info(:init_state, _state) do
-    {:noreply, read_data()}
+  @spec timezone_at(lng :: number, lat :: number) :: String.t() | nil
+  def timezone_at(lng, lat) when is_lng(lng) and is_lat(lat) do
+    point = %Geo.Point{coordinates: {lng, lat}}
+    GenServer.call(__MODULE__, {:timezone_at, point}, @timeout)
   end
 
-  defp read_data do
-    {:ok, [data_file]} = :zip.unzip(to_charlist(@data_archive))
-    data_file
-    |> File.read!()
-    |> :erlang.binary_to_term()
+  @doc """
+  Reload the timezone geo JSON data.
+
+  This allows for th data to be reloaded,
+  typically with a new release, without
+  restarting the application.
+
+  """
+  @spec reload_timezone_data :: :ok
+  def reload_timezone_data do
+    GenServer.call(__MODULE__, :reload_data, @timeout)
+  end
+
+  # --- Server callback implementation
+
+  def handle_continue(:load_data, _state) do
+    {:noreply, GeoData.load_compressed_data()}
+  end
+
+  def handle_call(:reload_data, _from, _state) do
+    case GeoData.load_compressed_data() do
+      {:ok, _data} = return -> {:reply, :ok, return}
+      other -> {:reply, other, other}
+    end
+  end
+
+  def handle_call(:version, _from, state) do
+    case state do
+      {:ok, [version | _tz_data]} -> {:reply, {:ok, version}, state}
+      other -> {:reply, other, state}
+    end
   end
 
   def handle_call({:timezone_at, %Geo.Point{} = point}, _from, state) do
     timezone =
-      state
-      |> Enum.filter(fn({_, _, bounding_box}) -> contains?(bounding_box, point) end)
-      |> Enum.find(fn({_, geometry, _}) -> contains?(geometry, point) end)
-      |> case do
-        {timezone, _, _} ->
-          timezone
-        nil ->
-          nil
+      with {:ok, [_version | tz_data]} <- state do
+        tz_data
+        |> Enum.filter(fn geometry -> contains?(geometry.properties.bounding_box, point) end)
+        |> Enum.find(&contains?(&1, point))
+        |> case do
+          %Geo.MultiPolygon{properties: %{tzid: tzid}} -> {:ok, tzid}
+          %Geo.Polygon{properties: %{tzid: tzid}} -> {:ok, tzid}
+          nil -> {:error, :notfound}
+        end
+      else
+        other -> other
```

Comment:
I'm not sure that I follow what `other` can be in this case.

Reply:
Yes, its redundant - `with` will return a default `else` clause with whatever the error is anyway. The error here could be that there is no data file and hence an `{:error, :noent}` may be returned. That is, `state` either holds the data, or the error noting that the data doesn't exist.

Reply:
I'm not sure that I understand why that is. If I understand correctly, that means that when the system starts, if it fails to download the latest release of the data, all subsequent calls to `TzWorld.timezone_at` will return `{:error, noent}`, which can only be fixed by running `mix  tz_world.update`. That means that users of the library need to have some kind of monitoring tool to be made aware that there is a problem.

I would have imagined that failing to download the data (or processing it for some reason, like available disk space for instance) would lead to the gen server to crashing, making the supervisor do its thing.

Did I miss something?

Reply:
At it currently stands, the data is not automatically downloaded, it has to be invoked via `mix tz_world.update` or by `TzWorld.Downloader.update_release/0`.  My thinking was that there may be situations - perhaps in an embedded nerves device, where assuming there's a network connection available at startup (or ever) would be unreasonable.

So I separated updating data from application start. I appreciate thats a design choice and perhaps its better to at least attempt to download the data when the app starts if it isn't already there (which is also what `tzdata` does).

WDYT?

Reply:
Oh yeah that makes sense. I was too hasty when reading the code, sorry. I was actually a bit worried about assuming a good network connection at startup, I think that your design is safer. Maybe the Readme should state that the app will not work properly until the update task is run (it says that you should run it but it's not explicit about what happens if you don't).

I guess that it will not be the best UX as a library user, since it's not "plug and play", but it does give you enough flexibility to make it work in any environment (I still wonder how you could share the data folder between deploys on something like kubernetes but I'm pretty sure there's a way).

> perhaps its better to at least attempt to download the data when the app starts if it isn't already there

It might be, but if you don't need it, and I don't, then I would rather wait until someone does before adding some more code.